report/report.tex

```diff
@@ -37,6 +66,50 @@ \section{Methodology}
+	
+	\paragraph{Definition of operators}
+
+	In \textit{Hanus}, operators can be defined to manipulate variables.  These operators always come in pairs, one operator for the forward manipulation and one operator for the reverse manipulation. There is no restriction that these operators should be different. For example, the reverse of the negation operator is the negation operator. However, in most circumstances, there operator are two different manipulations. In the original \textit{Janus}, there were 4 manipulation operators defined. There are the operators \textit{+=}, \textit{-=}, \textasciicircum$=$ and $\Leftrightarrow$. All of these operators only work on integers, as the original \textit{Janus} does not support any other types. The semantics are as follows:
+	\begin{itemize}
+		\item $x+=e$ adds any arbitrary expression $e$ to the variable $x$. The reverse of this statement is $x-=e$. As long as the variable on the right hand side of the operator is not used in the expression $e$, $e$ can by any computation that results in an integer. The expression $x+=x$ is therefore not allowed, because this statement does not have a reverse. The constraint that the variable on the left can not occur on the right applies for all operators.
+		\item $x-=e$ subtract the arbitrary expression $e$ from the variable $x$. The reverse is the statement $x+=e$.
+		\item $x$\textasciicircum$=e$ does a bitwise $xor$ with the expression $e$. This operator is its own reverse. 
+		\item $x\Leftrightarrow y$ is the swap manipulator. This operator takes two variables and swaps their values. This operator is its own reverse. 
+	\end{itemize}
+	For a user to implement their own operator in \textit{Hanus}, they have to define their operator to be the of the type \textit{Operator}. The type \textit{Operator} is defined in \textit{Haskell} as $Operator$ $\alpha$ $\beta = Operator$ $ (\alpha \rightarrow  \beta -\rightarrow  \alpha)$ $(\alpha \rightarrow  \beta \rightarrow \alpha)$. The first variable is the operator to be used in forward operation, the second variable is the reverse of the first operator. The definition of $+=$ in \textit{Hanus} is \textit{(+=) = Operator  (+) (-)}. For the convenience of the user, the function \textit{inverseOf} is provided. This function takes an \textit{Operator} and a function as its arguments. This functions returns an operator that is the reverse of the first parameter and uses the second operator in forward execution. $-=$ is defined using $inverseOf$ as $-= $  \textit{\;= inverseOf (+=) (-)}. Other operators that are defined for the users convenience include \textasciicircum$=$, swap, push and pop on a stack. All these operators are defined in \textit{Haskell}. Any user wanting to implement their own operators have to define these in \textit{Haskell}, as their is no way to define these in \textit{Hanus}.
```

Comment:
I removed `inverseOf` as it gave some problems with stack overflows. The new way to define operators is to first define an operator with both the forward and backward function, and define the reverse as `(-=) = inverse (+=)`.